### DISCUSSION TITLE
Closest equivalent for `git describe --tags`?

### DISCUSSION THREAD

**Discussion by BatmanAoD** [jj-vcs/jj#2563] [Q&A]
I really like using `git describe --tags` as an easy way to generate a version string. It has the following advantages:

* Extremely simple
* Always uniquely identifies the state of the code (as long as versions are not reassigned)
* If the current `HEAD` is on a commit corresponding to a release tag, you get that release version
* The commit-info is formatted as a valid semver prerelease tag; thus, if the version tags used are semver-compatible, the generated string is always semver-compatible

Is there anything like this in `jj`? For collocated repos `git describe --tags` works, but I think it would be good to have a native command that behaves similarly.

**Comment by moderation** [comment]
I had a similar question and just found https://github.com/hymkor/jjtagdesc. It doesn't look like it works anymore probably due to output changes in `jj` but it might provide some ideas

**Reply by shanesveller** [reply]
Loosely borrowed a few small particulars from this, and I came up with:
```bash
jj log -r 'latest(tags())::@- ~ empty()' --no-graph --reversed -T 'commit_id.short(10) ++ " " ++ tags ++ "\n"' \
| awk '{latest = $1; count++}; $2 != "" && tag == "" { tag = $2 } END {print tag "-" count "-" latest}'
```
My awk skill level is atrocious, so it can probably be golfed down some more. Tested with jj 0.25.0.

**Reply by lgarron** [reply]
@shanesveller: Thanks, that was very helpful! I found three discrepancies:

- Git includes a `g` at the beggining of the hash.
- Git includes 8 characters of the hash.
- This snippet counts an empty working commit, so the count is off. It's possible to use `(count - 1)`, but that will presumably be off if you haven't done `jj new` on a commit.

Fixing the first two gives me:

```shell
jj log -r 'latest(tags())::@- ~ empty()' --no-graph --reversed -T 'commit_id.short(8) ++ " " ++ tags ++ "\n"' \
  | awk '{latest = $1; count++}; $2 != "" && tag == "" { tag = $2 } END {print tag "-" count "-g" latest}'
```

**Comment by moderation** [comment]
This is awesome. Thanks for sharing. I converted to a Fish shell function which is working well:
```fish
# https://github.com/jj-vcs/jj/discussions/2563
function jj_describe --description 'replicate git describe --tags for jj'
    
    set count 0
    set tag ""
    set latest_commit ""

    set jj_output (jj log -r 'latest(tags())::@- ~ empty()' --no-graph --reversed -T 'commit_id.short(10) ++ " " ++ tags ++ "\n"')

    for line in $jj_output
        set count (math $count + 1)
        set -l parts (string split " " $line)
        set latest_commit $parts[1]

        if test (count $parts) -gt 1; and test -z "$tag"
            set tag (string sub --start 2 $parts[2])
        end
    end

    printf "%s-%d-%s\n" $tag $count $latest_commit

end
```

**Comment by lgarron** [comment]
I have a program that helps me implement common operations across repositories, across both `git` and `jj`: https://github.com/lgarron/repo

Matching `git describe --tags` is really not very simple.

- When the latest commit is a tag, the count and the hash should not be printed. Not too hard.
- I only just learned about `git_head()`. No one's mentioned it here yet!
- Counting commits must be done manually, with a chance for an off-by-one error depending on the implementation.
- https://git-scm.com/docs/git-describe says the number of hex digits in the hash "(will vary according to the number of objects in the repository with a default of 7) … or as many digits as needed to form a unique object name". It's not yet clear to me if [`.short()`](https://jj-vcs.github.io/jj/latest/templates/#operationid-type) implements the latter.
- When the minimum distance to each of multiple commits is the same, `jj`'s selection of the `latest()` is incorrect. https://git-scm.com/docs/git-describe states:

> If multiple tags were found during the walk then the tag which has the fewest commits different from the input commit-ish will be selected and output. Here fewest commits different is defined as the number of commits which would be shown by `git log tag..input` will be the smallest number of commits possible.

I don't see a way to do this with `jj` without listing all the closest tags and calculating the range for each. This seems silly and error-prone.[^1] Something like a `nearest()` function to match `git` would be useful here.

And that's all not even mentioning other flags for `git describe`, or even the functionality when `--tags` is not passed (which only considers annotated tags).

[^1]: In fact, I haven't even figured out how to list all closest tags. `heads(tags())` only returns a single result even if there are multiple tags at the same distance. Enumerating *all* tags would be unacceptably inefficient. I probably need something like `tags() & ~(tags()+..)`, except that doesn't work.

**Comment by jkuebart** [comment]
I know this is lame, but given the original motivation
> For collocated repos git describe --tags works

this worked for me as a very quick hack both for colocated and non-colocated repos:
```
git -C "$(jj git root)" \
    describe --tags \
    "$(jj log --limit 1 --no-graph --template commit_id)"
```

**Reply by bnjmnt4n** [reply]
You might want to use `jj git root`, in case the internal repository format changes.

**Reply by jkuebart** [reply]
Thanks! Great suggestion, updated.
